{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write a function GreedyMotifSearchWithPseudocounts(Dna, k, t) that takes a list of strings Dna followed by integers k and t and returns the result of running GreedyMotifSearch, where each profile matrix is generated with pseudocounts. Then add this function to Motifs.py. (Hint: Ideally, you should only need an extremely small modification to your original GreedyMotifSearch function.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input:  String Text and profile matrix Profile\n",
    "# Output: Pr(Text, Profile)\n",
    "def Pr(Motif, Profile):\n",
    "    prob=1.0\n",
    "    for char in range(len(Motif)):\n",
    "        prob *= Profile[Motif[char]][char] # What is the probability that this character is in this position?\n",
    "    \n",
    "    return prob\n",
    "\n",
    "def Score(Motifs):\n",
    "    #This really seems inefficient. We could have consensus return both values...\n",
    "    counts = CountWithPseudocounts(Motifs)\n",
    "    consensus, counts = Consensus(Motifs)\n",
    "    score = 0\n",
    "    for char in range(len(consensus)):\n",
    "        nucleotide = consensus[char] #Our nucleoutide\n",
    "        keys = [key for key in ['A','C',\"G\",\"T\"] if key != nucleotide ] # What are the non-consensus nucleotides\n",
    "        for key in keys:# For each key\n",
    "            score += counts[key][char] # add the number of times we were incorrect to our score\n",
    "    return score\n",
    "\n",
    "# Input:  A set of kmers Motifs\n",
    "# Output: A consensus string of Motifs.\n",
    "def Consensus(Motifs):\n",
    "    consensus = \"\" #empty string.\n",
    "    counts = CountWithPseudocounts(Motifs)\n",
    "\n",
    "    for j in range(len(Motifs[0])):\n",
    "        m = 0\n",
    "        frequentSymbol = \"\"\n",
    "        for symbol in \"ACGT\":\n",
    "            if counts[symbol][j] > m:\n",
    "                m = counts[symbol][j]\n",
    "                frequentSymbol = symbol\n",
    "        consensus += frequentSymbol # Add most frequent symbol\n",
    "    return consensus, counts\n",
    "\n",
    "# Input:  A set of kmers Motifs\n",
    "# Output: CountWithPseudocounts(Motifs)\n",
    "def CountWithPseudocounts(Motifs):\n",
    "    count = {} # initializing the count dictionary\n",
    "    #Initialize each nucleotide with an empty list, \n",
    "    for nucleotide in [\"A\",\"C\",\"G\",\"T\"]:\n",
    "        count[nucleotide] = []     \n",
    "    for ind in range(len(Motifs[0])):\n",
    "        for nucleotide in [\"A\",\"C\",\"G\",\"T\"]:\n",
    "            count[nucleotide].append(1.0) #everything must have a 1 initially  (Now that we are dealing with pseudo counts)\n",
    "        for motif in range(len(Motifs)): #For each Motif, loop through chars\n",
    "            count[Motifs[motif][ind]][ind] += 1.0 # FOr each nuc, increment its count for that Motif\n",
    "    return count\n",
    "\n",
    "# Write your ProfileMostProbableKmer() function here.\n",
    "# The profile matrix assumes that the first row corresponds to A, the second corresponds to C,\n",
    "# the third corresponds to G, and the fourth corresponds to T.\n",
    "# You should represent the profile matrix as a dictionary whose keys are 'A', 'C', 'G', and 'T' and whose values are lists of floats\n",
    "def ProfileMostProbableKmer(text, k, Profile):\n",
    "    most_prob = \"\"\n",
    "    high_prob = -1.0\n",
    "    for index in range(len(text)-k+1):\n",
    "        Motif = text[index:index+k]\n",
    "        Motif_prob = Pr(Motif, Profile)\n",
    "        if Motif_prob > high_prob:\n",
    "            high_prob = Motif_prob\n",
    "            most_prob = Motif\n",
    "    return most_prob\n",
    "\n",
    "# Input:  A set of kmers Motifs\n",
    "# Output: ProfileWithPseudocounts(Motifs)\n",
    "def ProfileWithPseudocounts(Motifs):\n",
    "    t = len(Motifs)\n",
    "    k = len(Motifs[0])\n",
    "    profile = {}\n",
    "    counts = CountWithPseudocounts(Motifs)\n",
    "\n",
    "    for nucleotide in [\"A\",\"C\",\"G\",\"T\"]:\n",
    "        #everything must divided by the total number of Motifs  \n",
    "        #We added a single count to each nucleotide, hence the +4.0\n",
    "        profile[nucleotide] = [ float(count)/(float(t) + 4.0) for count in counts[nucleotide]]  \n",
    "    return profile\n",
    "\n",
    "def GreedyMotifSearchWithPseudocounts(Dna, k, t):\n",
    "    BestMotifs = []\n",
    "    for i in range(0, t):\n",
    "        BestMotifs.append(Dna[i][0:k])\n",
    "        \n",
    "    n = len(Dna[0])\n",
    "    for i in range(n-k+1):\n",
    "        Motifs = []\n",
    "        Motifs.append(Dna[0][i:i+k])\n",
    "        for j in range(1, t):\n",
    "            P = ProfileWithPseudocounts(Motifs[0:j])\n",
    "            Motifs.append(ProfileMostProbableKmer(Dna[j], k, P))\n",
    "        if Score(Motifs) < Score(BestMotifs):\n",
    "            BestMotifs = Motifs\n",
    "    return BestMotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Test 0 # Sample Dataset (your code is not run on this dataset)\n",
    "Input:\"\"\"\n",
    "\n",
    "\"\"\"    3 5\n",
    "    GGCGTTCAGGCA\n",
    "    AAGAATCAGTCA\n",
    "    CAAGGAGTTCGC\n",
    "    CACGTCAATCAC\n",
    "    CAATAATATTCG\n",
    "Output:\n",
    "    TTC\n",
    "    ATC\n",
    "    TTC\n",
    "    ATC\n",
    "    TTC\n",
    "\n",
    "Test 1 # Check if you're missing the first kmer in each string of Dna\n",
    "Input\n",
    "    5 8\n",
    "    AGGCGGCACATCATTATCGATAACGATTCGCCGCATTGCC\n",
    "    ATCCGTCATCGAATAACTGACACCTGCTCTGGCACCGCTC\n",
    "    AAGCGTCGGCGGTATAGCCAGATAGTGCCAATAATTTCCT\n",
    "    AGTCGGTGGTGAAGTGTGGGTTATGGGGAAAGGCAGACTG\n",
    "    AACCGGACGGCAACTACGGTTACAACGCAGCAAGAATATT\n",
    "    AGGCGTCTGTTGTTGCTAACACCGTTAAGCGACGGCAACT\n",
    "    AAGCGGCCAACGTAGGCGCGGCTTGGCATCTCGGTGTGTG\n",
    "    AATTGAAAGGCGCATCTTACTCTTTTCGCTTTCAAAAAAA\n",
    "Output:\n",
    "    AGGCG\n",
    "    ATCCG\n",
    "    AAGCG\n",
    "    AGTCG\n",
    "    AACCG\n",
    "    AGGCG\n",
    "    AGGCG\n",
    "    AGGCG\n",
    "\n",
    "Test 2 # Check if you're missing the last kmer in each string of Dna\n",
    "Input\n",
    "    5 8\n",
    "    GCACATCATTAAACGATTCGCCGCATTGCCTCGATAGGCG\n",
    "    TCATAACTGACACCTGCTCTGGCACCGCTCATCCGTCGAA\n",
    "    AAGCGGGTATAGCCAGATAGTGCCAATAATTTCCTTCGGC\n",
    "    AGTCGGTGGTGAAGTGTGGGTTATGGGGAAAGGCAGACTG\n",
    "    AACCGGACGGCAACTACGGTTACAACGCAGCAAGAATATT\n",
    "    AGGCGTCTGTTGTTGCTAACACCGTTAAGCGACGGCAACT\n",
    "    AAGCTTCCAACATCGTCTTGGCATCTCGGTGTGTGAGGCG\n",
    "    AATTGAACATCTTACTCTTTTCGCTTTCAAAAAAAAGGCG\n",
    "Output:\n",
    "    AGGCG\n",
    "    TGGCA\n",
    "    AAGCG\n",
    "    AGGCA\n",
    "    CGGCA\n",
    "    AGGCG\n",
    "    AGGCG\n",
    "    AGGCG\n",
    "\n",
    "Test 3 # Check if you're not breaking ties correctly (you should pick the first kmer in the case of a tie)\n",
    "Input\n",
    "    5 8\n",
    "    GCACATCATTATCGATAACGATTCATTGCCAGGCGGCCGC\n",
    "    TCATCGAATAACTGACACCTGCTCTGGCTCATCCGACCGC\n",
    "    TCGGCGGTATAGCCAGATAGTGCCAATAATTTCCTAAGCG\n",
    "    GTGGTGAAGTGTGGGTTATGGGGAAAGGCAGACTGAGTCG\n",
    "    GACGGCAACTACGGTTACAACGCAGCAAGAATATTAACCG\n",
    "    TCTGTTGTTGCTAACACCGTTAAGCGACGGCAACTAGGCG\n",
    "    GCCAACGTAGGCGCGGCTTGGCATCTCGGTGTGTGAAGCG\n",
    "    AAAGGCGCATCTTACTCTTTTCGCTTTCAAAAAAAAATTG\n",
    "Output:\n",
    "    GGCGG\n",
    "    GGCTC\n",
    "    GGCGG\n",
    "    GGCAG\n",
    "    GACGG\n",
    "    GACGG\n",
    "    GGCGC\n",
    "    GGCGC\n",
    "\n",
    "Test 4 # Full dataset\n",
    "Input:\"\"\"\n",
    "k,t =3, 8\n",
    "Dna4= ['GCACATCATTATCGATAACGATTCATTGCCAGGCGGCCGC',\n",
    "    'TCATCGAATAACTGACACCTGCTCTGGCTCATCCGACCGC',\n",
    "    'TCGGCGGTATAGCCAGATAGTGCCAATAATTTCCTAAGCG',\n",
    "    'GTGGTGAAGTGTGGGTTATGGGGAAAGGCAGACTGAGTCG',\n",
    "    'GACGGCAACTACGGTTACAACGCAGCAAGAATATTAACCG',\n",
    "    'TCTGTTGTTGCTAACACCGTTAAGCGACGGCAACTAGGCG',\n",
    "    'GCCAACGTAGGCGCGGCTTGGCATCTCGGTGTGTGAAGCG',\n",
    "    'AAAGGCGCATCTTACTCTTTTCGCTTTCAAAAAAAAATTG']\n",
    "#Output:\n",
    "out4=    [\"GGC\",\n",
    "    \"GGC\",\n",
    "    \"GGC\",\n",
    "    \"GGC\",\n",
    "    \"GGC\",\n",
    "    \"GGC\",\n",
    "    \"GGC\",\n",
    "    \"GGC\"]\n",
    "assert (GreedyMotifSearch(Dna4, k, t) == out4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GGC', 'GGC', 'GGC', 'GGC', 'GGC', 'GGC', 'GGC', 'GGC']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GreedyMotifSearch(Dna4, k, t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "Dna = [\"GCGCCCCGCCCGGACAGCCATGCGCTAACCCTGGCTTCGATGGCGCCGGCTCAGTTAGGGCCGGAAGTCCCCAATGTGGCAGACCTTTCGCCCCTGGCGGACGAATGACCCCAGTGGCCGGGACTTCAGGCCCTATCGGAGGGCTCCGGCGCGGTGGTCGGATTTGTCTGTGGAGGTTACACCCCAATCGCAAGGATGCATTATGACCAGCGAGCTGAGCCTGGTCGCCACTGGAAAGGGGAGCAACATC\",\n",
    "\"CCGATCGGCATCACTATCGGTCCTGCGGCCGCCCATAGCGCTATATCCGGCTGGTGAAATCAATTGACAACCTTCGACTTTGAGGTGGCCTACGGCGAGGACAAGCCAGGCAAGCCAGCTGCCTCAACGCGCGCCAGTACGGGTCCATCGACCCGCGGCCCACGGGTCAAACGACCCTAGTGTTCGCTACGACGTGGTCGTACCTTCGGCAGCAGATCAGCAATAGCACCCCGACTCGAGGAGGATCCCG\",\n",
    "\"ACCGTCGATGTGCCCGGTCGCGCCGCGTCCACCTCGGTCATCGACCCCACGATGAGGACGCCATCGGCCGCGACCAAGCCCCGTGAAACTCTGACGGCGTGCTGGCCGGGCTGCGGCACCTGATCACCTTAGGGCACTTGGGCCACCACAACGGGCCGCCGGTCTCGACAGTGGCCACCACCACACAGGTGACTTCCGGCGGGACGTAAGTCCCTAACGCGTCGTTCCGCACGCGGTTAGCTTTGCTGCC\",\n",
    "\"GGGTCAGGTATATTTATCGCACACTTGGGCACATGACACACAAGCGCCAGAATCCCGGACCGAACCGAGCACCGTGGGTGGGCAGCCTCCATACAGCGATGACCTGATCGATCATCGGCCAGGGCGCCGGGCTTCCAACCGTGGCCGTCTCAGTACCCAGCCTCATTGACCCTTCGACGCATCCACTGCGCGTAAGTCGGCTCAACCCTTTCAAACCGCTGGATTACCGACCGCAGAAAGGGGGCAGGAC\",\n",
    "\"GTAGGTCAAACCGGGTGTACATACCCGCTCAATCGCCCAGCACTTCGGGCAGATCACCGGGTTTCCCCGGTATCACCAATACTGCCACCAAACACAGCAGGCGGGAAGGGGCGAAAGTCCCTTATCCGACAATAAAACTTCGCTTGTTCGACGCCCGGTTCACCCGATATGCACGGCGCCCAGCCATTCGTGACCGACGTCCCCAGCCCCAAGGCCGAACGACCCTAGGAGCCACGAGCAATTCACAGCG\",\n",
    "\"CCGCTGGCGACGCTGTTCGCCGGCAGCGTGCGTGACGACTTCGAGCTGCCCGACTACACCTGGTGACCACCGCCGACGGGCACCTCTCCGCCAGGTAGGCACGGTTTGTCGCCGGCAATGTGACCTTTGGGCGCGGTCTTGAGGACCTTCGGCCCCACCCACGAGGCCGCCGCCGGCCGATCGTATGACGTGCAATGTACGCCATAGGGTGCGTGTTACGGCGATTACCTGAAGGCGGCGGTGGTCCGGA\",\n",
    "\"GGCCAACTGCACCGCGCTCTTGATGACATCGGTGGTCACCATGGTGTCCGGCATGATCAACCTCCGCTGTTCGATATCACCCCGATCTTTCTGAACGGCGGTTGGCAGACAACAGGGTCAATGGTCCCCAAGTGGATCACCGACGGGCGCGGACAAATGGCCCGCGCTTCGGGGACTTCTGTCCCTAGCCCTGGCCACGATGGGCTGGTCGGATCAAAGGCATCCGTTTCCATCGATTAGGAGGCATCAA\",\n",
    "\"GTACATGTCCAGAGCGAGCCTCAGCTTCTGCGCAGCGACGGAAACTGCCACACTCAAAGCCTACTGGGCGCACGTGTGGCAACGAGTCGATCCACACGAAATGCCGCCGTTGGGCCGCGGACTAGCCGAATTTTCCGGGTGGTGACACAGCCCACATTTGGCATGGGACTTTCGGCCCTGTCCGCGTCCGTGTCGGCCAGACAAGCTTTGGGCATTGGCCACAATCGGGCCACAATCGAAAGCCGAGCAG\",\n",
    "\"GGCAGCTGTCGGCAACTGTAAGCCATTTCTGGGACTTTGCTGTGAAAAGCTGGGCGATGGTTGTGGACCTGGACGAGCCACCCGTGCGATAGGTGAGATTCATTCTCGCCCTGACGGGTTGCGTCTGTCATCGGTCGATAAGGACTAACGGCCCTCAGGTGGGGACCAACGCCCCTGGGAGATAGCGGTCCCCGCCAGTAACGTACCGCTGAACCGACGGGATGTATCCGCCCCAGCGAAGGAGACGGCG\",\n",
    "\"TCAGCACCATGACCGCCTGGCCACCAATCGCCCGTAACAAGCGGGACGTCCGCGACGACGCGTGCGCTAGCGCCGTGGCGGTGACAACGACCAGATATGGTCCGAGCACGCGGGCGAACCTCGTGTTCTGGCCTCGGCCAGTTGTGTAGAGCTCATCGCTGTCATCGAGCGATATCCGACCACTGATCCAAGTCGGGGGCTCTGGGGACCGAAGTCCCCGGGCTCGGAGCTATCGGACCTCACGATCACC\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GGACTTCAGGCCCTA', 'GGTCAAACGACCCTA', 'GGACGTAAGTCCCTA', 'GGATTACCGACCGCA', 'GGCCGAACGACCCTA', 'GGACCTTCGGCCCCA', 'GGACTTCTGTCCCTA', 'GGACTTTCGGCCCTG', 'GGACTAACGGCCCTC', 'GGACCGAAGTCCCCG']\n",
      "80.0\n"
     ]
    }
   ],
   "source": [
    "\"\"\"Exercise Break (1 point): Apply GreedyMotifSearch with pseudocounts to find motifs in the DosR \n",
    "dataset with k-mer length equal to 15 (click here to download).\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "# set t equal to the number of strings in Dna and k equal to 15\n",
    "k, t = 15, 10\n",
    "# Call GreedyMotifSearchWithPseudocounts(Dna, k, t) and store the output in a variable called Motifs\n",
    "Motifs = GreedyMotifSearchWithPseudocounts(Dna, k, t)\n",
    "# Print the Motifs variable\n",
    "print (Motifs)\n",
    "# Print Score(Motifs)\n",
    "print (Score(Motifs))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You have now seen the power of pseudocounts illustrated on a small example. Running GreedyMotifSearch with pseudocounts to solve the Subtle Motif Problem returns a collection of 15-mers Motifs with Score(Motifs) = 41 and Consensus(Motifs) = \"AAAAAtAgaGGGGtt\". Thus, Laplace’s Rule of Succession has provided a significant improvement over the original GreedyMotifSearch, which returned the consensus string \"gttAAAtAgaGatGtG\" with Score(Motifs) = 58.\n",
    "\n",
    "You may be satisfied with the performance of GreedyMotifSearch, but you should know by now that your authors are never satisfied. Can we design an even more accurate motif finding algorithm?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
