{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Many algorithms are iterative procedures that must choose among various alternatives at each iteration. Some of these alternatives may lead to correct solutions, whereas others may not. Greedy algorithms select the “most attractive” alternative at each iteration. For example, a greedy algorithm in chess might attempt to capture an opponent’s most valuable piece at every move. Yet anyone who has played chess knows that a strategy looking only one move ahead will likely produce disastrous results.\n",
    "\n",
    "In general, most greedy algorithms typically fail to find an exact solution of the problem; instead, they are often fast heuristics that trade accuracy for speed in order to find an approximate solution. Nevertheless, for many biological problems, greedy algorithms will prove quite useful."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this section, we will explore a greedy approach to motif finding. Again, let Motifs be a collection of k-mers taken from t strings Dna. We can view each column of Profile(Motifs) as a four-sided biased die with one nucleotide on each side. Thus, a profile matrix with k columns can be viewed as a collection of k dice that we will roll to randomly generate a k-mer. For example, if the first column of the profile matrix is (0.2, 0.1, 0.0, 0.7), then we generate A as the first nucleotide with probability 0.2, C with probability 0.1, G with probability 0.0, and T with probability 0.7.\n",
    "\n",
    "In the figure below, we reproduce the profile matrix for the NF-kB binding sites, where the lone colored entry in the i-th column corresponds to the i-th nucleotide in \"ACGGGGATTACC\". The probability Pr(\"ACGGGGATTACC\", Profile) that Profile generates \"ACGGGGATTACC\" is computed by multiplying the highlighted entries in the profile matrix.\n",
    "\n",
    "\n",
    "\n",
    "Figure: Generating a random string based on a profile matrix by selecting the i-th nucleotide in the string with the probability corresponding to that nucleotide in the i-th column of the profile matrix. The probability that a profile matrix will produce a given string is given by the product of individual nucleotide probabilities."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " Generating a random string based on a profile matrix by selecting the i-th nucleotide in the string with the probability corresponding to that nucleotide in the i-th column of the profile matrix. The probability that a profile matrix will produce a given string is given by the product of individual nucleotide probabilities."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A k-mer tends to have a higher probability when it is more similar to the consensus string of a profile. For example, for the NF-κB profile matrix (reproduced at bottom) and its consensus string \"TCGGGGATTTCC\",\n",
    "\n",
    "Pr(\"TCGGGGATTTCC\", Profile) = .7 · .6 · 1 · 1 · .9 · .9 · .9 · .5 · .8 · .7 · .4 · .6 = 0.0205753 ,\n",
    "\n",
    "which is larger than the value of Pr(\"ACGGGGATTACC\", Profile) = 0.000839808 that we computed on the previous step.\n",
    "\n",
    "Exercise Break (1 point): Compute Pr(\"TCGTGGATTTCC\", Profile) for the NF-κB profile matrix shown below. (Note: you can do this exercise with a calculator.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input:  String Text and profile matrix Profile\n",
    "# Output: Pr(Text, Profile)\n",
    "def Pr(Motif, Profile):\n",
    "    prob=1.0\n",
    "    for char in range(len(Motif)):\n",
    "        prob *= Profile[Motif[char]][char] # What is the probability that this character is in this position?\n",
    "    \n",
    "    return prob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Profile:\n",
    "Profile = {\n",
    "    'A': [0.2, 0.2, 0.0, 0.0, 0.0, 0.0, 0.9, 0.1, 0.1, 0.1, 0.3, 0.0],\n",
    "    'C': [0.1, 0.6, 0.0, 0.0, 0.0, 0.0, 0.0, 0.4, 0.1, 0.2, 0.4, 0.6],\n",
    "    'G': [0.0, 0.0, 1.0, 1.0, 0.9, 0.9, 0.1, 0.0, 0.0, 0.0, 0.0, 0.0],\n",
    "    'T': [0.7, 0.2, 0.0, 0.0, 0.1, 0.1, 0.0, 0.5, 0.8, 0.7, 0.3, 0.4]\n",
    "}\n",
    "\n",
    "Motif = \"TCGTGGATTTCC\"\n",
    "Pr(Motif, Profile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given a profile matrix Profile, we can compute the probability of every k-mer in a string Text and find a Profile-most probable k-mer in Text, i.e., a k-mer that was most likely to have been generated by Profile among all k-mers in Text. For the NF-κB profile matrix, \"ACGGGGATTACC\" is the Profile-most probable 12-mer in \"ggtACGGGGATTACCt\". Indeed, every other 12-mer in this string has probability 0. In general, if there are multiple Profile-most probable k-mers in Text, then we select the first such k-mer occurring in Text.\n",
    "```\n",
    "Profile-most Probable k-mer Problem: Find a Profile-most probable k-mer in a string. \n",
    " Input: A string Text, an integer k, and a 4 x k matrix Profile.\n",
    " Output: A Profile-most probable k-mer in Text.\n",
    "```\n",
    "Code Challenge (3 points): Solve the Profile-most Probable k-mer Problem by writing a function ProfileMostProbablePattern(Text, k, Profile). (Hint: make sure to use the function Pr(Text, Profile) as a subroutine.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write your ProfileMostProbableKmer() function here.\n",
    "# The profile matrix assumes that the first row corresponds to A, the second corresponds to C,\n",
    "# the third corresponds to G, and the fourth corresponds to T.\n",
    "# You should represent the profile matrix as a dictionary whose keys are 'A', 'C', 'G', and 'T' and whose values are lists of floats\n",
    "def ProfileMostProbableKmer(text, k, Profile):\n",
    "    most_prob = \"\"\n",
    "    high_prob = -1.0\n",
    "    for index in range(len(text)-k+1):\n",
    "        Motif = text[index:index+k]\n",
    "        Motif_prob = Pr(Motif, Profile)\n",
    "        if Motif_prob > high_prob:\n",
    "            high_prob = Motif_prob\n",
    "            most_prob = Motif\n",
    "    return most_prob\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ProfileTestCase0= { 'A': [0.2, 0.2, 0.3, 0.2, 0.3],'C': [0.4, 0.3, 0.1, 0.5, 0.1], 'G': [0.3, 0.3, 0.5, 0.2, 0.4], 'T': [0.1, 0.2, 0.1, 0.1, 0.2]}\n",
    "text0=    \"ACCTGTTTATTGCCTAAGTTCCGAACAAACCCAATATAGCCCGAGGGCCT\"\n",
    "k0=    5\n",
    "out0=\"CCGAG\"\n",
    "assert (ProfileMostProbableKmer(text0,k0,ProfileTestCase0)==out0)\n",
    "\n",
    "ProfileTestCase1= { 'A': [ 0.7, 0.2, 0.1, 0.5, 0.4, 0.3, 0.2, 0.1],'C': [ 0.2, 0.2, 0.5, 0.4, 0.2, 0.3, 0.1, 0.6], 'G': [0.1, 0.3, 0.2, 0.1, 0.2, 0.1, 0.4, 0.2], 'T': [ 0.0, 0.3, 0.2, 0.0, 0.2, 0.3, 0.3, 0.1]}\n",
    "text1=     \"AGCAGCTTTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATCTGAACTGGTTACCTGCCGTGAGTAAAT\"\n",
    "k1=     8\n",
    "out1=\"AGCAGCTT\"\n",
    "assert (ProfileMostProbableKmer(text1,k1,ProfileTestCase1)==out1)\n",
    "ProfileTestCase2= { 'A': [ 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.1, 0.2, 0.3, 0.4, 0.5],'C': [ 0.3, 0.2, 0.1, 0.1, 0.2, 0.1, 0.1, 0.4, 0.3, 0.2, 0.2, 0.1], 'G': [0.2, 0.1, 0.4, 0.3, 0.1, 0.1, 0.1, 0.3, 0.1, 0.1, 0.2, 0.1], 'T': [ 0.3, 0.4, 0.1, 0.1, 0.1, 0.1, 0.0, 0.2, 0.4, 0.4, 0.2, 0.3]}\n",
    "text2=     \"TTACCATGGGACCGCTGACTGATTTCTGGCGTCAGCGTGATGCTGGTGTGGATGACATTCCGGTGCGCTTTGTAAGCAGAGTTTA\"\n",
    "k2=     12\n",
    "out2=\"AAGCAGAGTTTA\"\n",
    "assert (ProfileMostProbableKmer(text2,k2,ProfileTestCase2)==out2)\n",
    "ProfileTestCase3= { 'A': [ 1.0 ,1.0, 1.0],'C': [ 0.0, 0.0, 0.0], 'G': [ 0.0, 0.0, 0.0], 'T': [ 0.0, 0.0, 0.0]}\n",
    "text3=     \"AACCGGTT\"\n",
    "k3=     3\n",
    "out3=\"AAC\"\n",
    "assert (ProfileMostProbableKmer(text3,k3,ProfileTestCase3)==out3)\n",
    "ProfileTestCase4= { 'A': [ 0.2, 0.2, 0.3, 0.2, 0.3], 'C': [ 0.4, 0.3, 0.1, 0.5, 0.1], 'G': [ 0.3, 0.3, 0.5, 0.2, 0.4], 'T': [0.1, 0.2, 0.1 ,0.1, 0.2]}\n",
    "text4=     \"TTACCATGGGACCGCTGACTGATTTCTGGCGTCAGCGTGATGCTGGTGTGGATGACATTCCGGTGCGCTTTGTAAGCAGAGTTTA\"\n",
    "k4=     5\n",
    "out4=\"CAGCG\"\n",
    "assert (ProfileMostProbableKmer(text4,k4,ProfileTestCase4)==out4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0054"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Quiz\n",
    "Profile = {\n",
    "    \"A\":[ 0.4, 0.3, 0.0, 0.1, 0.0, 0.9],\n",
    "\n",
    "\"C\": [0.2, 0.3, 0.0, 0.4, 0.0, 0.1],\n",
    "\n",
    "\"G\": [0.1, 0.3, 1.0, 0.1, 0.5, 0.0],\n",
    "\n",
    "\"T\":[ 0.3, 0.1, 0.0, 0.4, 0.5, 0.0]\n",
    "    \n",
    "}\n",
    "\n",
    "Pr(\"GAGCTA\",Profile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.00405"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Pr(\"TCGGTA\",Profile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
